File: milliseconds/golang/controllers/millisecond_controller.go
package controllers

import (
	"net/http"
	modelrequests "note-golang-millisecond/models/requests"
	"note-golang-millisecond/services"

	"github.com/labstack/echo/v4"
)

type MillisecondController interface {
	GetByGmtPlus8(c echo.Context) error
	GetByGmtMinus8(c echo.Context) error
}

type millisecondController struct {
	millisecondService services.MillisecondService
}

func NewMillisecondController(millisecondService services.MillisecondService) MillisecondController {
	return &millisecondController{
		millisecondService: millisecondService,
	}
}

func (controller *millisecondController) GetByGmtPlus8(c echo.Context) error {
	var millisecondRequest modelrequests.MillisecondRequest
	err := c.Bind(&millisecondRequest)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"response": err.Error(),
		})
	}
	response := controller.millisecondService.GetByGMTPlus8(c.Request().Context(), millisecondRequest)
	return c.JSON(http.StatusOK, map[string]interface{}{
		"response": response,
	})
}

func (controller *millisecondController) GetByGmtMinus8(c echo.Context) error {
	var millisecondRequest modelrequests.MillisecondRequest
	err := c.Bind(&millisecondRequest)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"response": err.Error(),
		})
	}
	response := controller.millisecondService.GetByGMTMinus8(c.Request().Context(), millisecondRequest)
	return c.JSON(http.StatusOK, map[string]interface{}{
		"response": response,
	})
}

File: milliseconds/rust/src/controllers/millisecond_controller.rs
use std::sync::Arc;

use axum::{extract::State, http::StatusCode, response::IntoResponse, Json};

use crate::{models::requests::millisecond_request::MillisecondRequest, services::millisecond_service::MillisecondService, states::app_state::AppState};

pub async fn get_by_gmmt_plus8(
    State(state): State<Arc<AppState>>,
    Json(millisecond_request): Json<MillisecondRequest>
) -> impl IntoResponse {
    let response = state.millisecond_service.get_by_gmt_plus8(millisecond_request).await;
    (StatusCode::OK, Json(response))
}

pub async fn get_by_gmt_minus8(
    State(state): State<Arc<AppState>>,
    Json(millisecond_request): Json<MillisecondRequest>
) -> impl IntoResponse {
    let response = state.millisecond_service.get_by_gmt_minus8(millisecond_request).await;
    (StatusCode::OK, Json(response))
}